keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,972 @@
+* [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This proposal aims to allow the pod status to express which pod updates are 
+currently being reflected in the pod status. The idea is to leverage the 
+existing `metadata.Generation` field and add a new `status.observedGeneration` field to 
+the pod status. 
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+One of the motivations for this KEP comes from the existing ResizeStatus field. 
+In its original implementation, the ResizeStatus field was written to by both 
+the API server and the Kubelet, creating a race condition on updates. Removing 
+the Proposed state makes the Kubelet the only writer to that field, but leaves a
+gap in knowing whether the latest resize has been acknowledged by the Kubelet. 
+The changes proposed in this KEP resolve this gap.
+
+The ResizeStatus is used as an example here, but in practice this issue can be 
+generally found in any type of pod update.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Provide a general solution for the pod status to express which pod update is 
+currently being reflected.
+
+### Non-Goals
```

Comment:
is any of this logic be applicable on static pods? Should we explicitly make them non-goal?

Reply:
(caveat: I'm new in this area of k8s so will need some guidance from experts on the right thing to do)

My two cents: I don't really see a need to populate generation or observedGeneration on static pods, so I'm happy to add that as a non-goal.

My open question: If we write generic code in the API server to populate / increment pod generation, does that mean mirror pods will have their generation field populated (and is that behavior desirable)? In a similar vein is their any value in populating observedGeneration on the mirror pod? 

cc @tallclair

Reply:
and will the kubelet code be generic or should exclude static pod's generation and observedGeneration.

Perhaps it may go to implementation details section instead of non-goals. But we need to decide what is the desireable behavior for those from API standpoint

Reply:
I agree with making static pods out-of-scope.

Reply:
we still need to describe what the expectations are from API standpoint:

1. Do we set generation to `1` on creation?
2. Do we accept any observedGeneration, do we drop them or ignore them?
3. Does kubelet have to make sure it never sends the observedGeneration for the mirror pod statuses?

Reply:
I am fine explicitly state generation for static pods is out-of-scope for this release. But for a long run, I think generation and observedGeneration are beneficial for the static pods too. I hope we can dive deep on it next release. cc/ @yujuhong for vis.

Reply:
what about resize? Do we include requests and limits into the UID generation? I rememebr there was a bug around it

Reply:
and we still need to describe what behavior and API checks we need to make on API side and how kubelet needs to set or not set this field on status updates in alpha. Making kubelet and API behaviors inconsistent for mirror pods in alpha will result in issues

Reply:
> and we still need to describe what behavior and API checks we need to make on API side and how kubelet needs to set or not set this field on status updates in alpha. Making kubelet and API behaviors inconsistent for mirror pods in alpha will result in issues

Here is my proposal on this for mirror pods:
1. API server never sets `metadata.generation` for mirror pods. 
2. If there is `metadata.generation` or `status.observedGeneration` set in the incoming config for a static pod, we ignore and drop them on the API side.
3. Kubelet never sets `status.observedGeneration` for mirror pods (it will need to explicitly exclude them).

Am I missing any cases?

Reply:
I'm not sure I'd do anything special on the apiserver side... if kubelet deletes / recreates mirror pods when the static pod manifest changes already, won't that effectively keep all static pods at generation 1?

Reply:
> won't that effectively keep all static pods at generation 1

I was under the impression that it would be preferable to keep the generation unset (generation=0), but if we want to have them be generation 1 then yeah the apiserver doesn't need to do anything special there

Reply:
I have added the proposal for the behavior and API checks of mirror pods to the body of the KEP. If there are concerns about that (such as whether they should have generation 1 or 0) please feel free to leave comments on it.

I also added to the "future enhancements" section that we may want to consider adding these fields on static pods for a future release. 

I'm going to mark this discussion as resolved but please feel free to reopen if appropriate.